src/parsing/abstract-parser.h

```diff
@@ -2004,41 +2004,89 @@ void AbstractParser<Impl>::ParseFunction(
   scanner_.Initialize();
 
   FunctionLiteral* result = nullptr;
-  if (V8_UNLIKELY(shared_info->HasUncompiledDataWithBinAstParseData())) {
-    Handle<UncompiledDataWithBinAstParseData> uncompiled_data = handle(shared_info->uncompiled_data_with_binast_parse_data(), isolate);
-    Handle<ByteArray> binast_parse_data = handle(uncompiled_data->binast_parse_data(), isolate);
-
-    RuntimeCallTimerScope runtime_timer(impl()->runtime_call_stats_,
-                                      RuntimeCallCounterId::kDeserializeBinAst);
+  if (V8_UNLIKELY(shared_info->HasUncompiledDataWithBinAstParseData() ||
+                  shared_info->HasUncompiledDataWithInnerBinAstParseData())) {
+    RuntimeCallTimerScope runtime_timer(
+        impl()->runtime_call_stats_, RuntimeCallCounterId::kDeserializeBinAst);
     auto start = std::chrono::high_resolution_clock::now();
+    bool is_inner = shared_info->HasUncompiledDataWithInnerBinAstParseData();
+    if (is_inner) {
```

Comment:
Hmm is there any way we could dedupe most of the code in these two branches, maybe through the use of Optional?

Reply:
yeah whoops, meant to rewrite this. after your change with the handles here i had some trouble keeping it not duped, so duplicated it and meant to figure it out later